src/transformers/models/llama/modeling_llama.py

```diff
@@ -352,18 +354,24 @@ def forward(
         key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
+        if isinstance(past_key_value, Cache):
+            past_key_value.update_pre_rotation(key_states, value_states)
```

Comment:
I'm very much not a fan of having an `isinstance` conditional every time, but it's probably the most convenient way to keep backwards compatibility.

Also, I'm considering renaming `update_pre_rotation` and `update` to `on_...` much like the `TrainerCallback`. I think that might be more extendible etc.
Note that `update_pre_rotation` only exists to eventually support the Attention Sink Cache.

Reply:
Same, I think it's fine to just assume that inside the attention layer the `past_key_value` is always of type `Cache` - only when returning `past_key_value` from the forward we have to transform it back no? Intermediate classes are not considered public API (they can't be imported from `__init__.py`). Wdyt @gante ?

Reply:
Tackled in 1129513b3fe3a540bbeebe4fcef8bc5b29903ad8.